### ISSUE TITLE
Images missing, repeats ignored, VAECache warnings - SD3 full finetuning

### ISSUE THREAD

**Issue by a-l-e-x-d-s-9** [bghira/SimpleTuner#484]
I'm testing SD3 full finetuning with the latest version of SimpleTuner.
The issues I see today:
1. I have 612 images in the dataset, they all are JPG files, but only 567 are used. Seems the script just deleted the aspect ratios that are below my batch size of 27. I don't see any option to change it so they would not be deleted. Using repeats could have helped to fill not full buckets and a few repetitions of files wouldn't be as bad as deleting the whole bucket.
My settings:
``` bash
# If this is set, any images that fail to open will be DELETED to avoid re-checking them every time.
export DELETE_ERRORED_IMAGES=1
# If this is set, any images that are too small for the minimum resolution size will be DELETED.
export DELETE_SMALL_IMAGES=0 
```

log:
``` log
3:43:44,728 [INFO] (DataBackendFactory) Completed loading text embed services.
2024-06-18 13:43:44,728 [INFO] (DataBackendFactory) Configuring data backend: all_dataset
2024-06-18 13:43:44,728 [INFO] (DataBackendFactory) Configured backend: {'id': 'all_dataset', 'config': {'vae_cache_clear_each_epoch': False, 'probability': 1.0, 'repeats': 5, 'crop': False, 'crop_aspect': 'preserve', 'crop_aspect_buckets': None, 'crop_style': 'random', 'disable_validation': False, 'resolution': 0.5, 'resolution_type': 'area', 'caption_strategy': 'textfile', 'instance_data_root': '/workspace/input/dataset', 'maximum_image_size': 1.0, 'target_downsample_size': 0.55}, 'dataset_type': 'image'}
2024-06-18 13:43:44,728 [INFO] (DataBackendFactory) (id=all_dataset) Loading bucket manager.
2024-06-18 13:43:44,733 [INFO] (JsonMetadataBackend) Checking for cache file: /workspace/input/dataset/aspect_ratio_bucket_indices.json
2024-06-18 13:43:44,733 [WARNING] (JsonMetadataBackend) No cache file found, creating new one.
2024-06-18 13:43:44,734 [INFO] (DataBackendFactory) (id=all_dataset) Refreshing aspect buckets on main process.
2024-06-18 13:43:44,734 [INFO] (BaseMetadataBackend) Discovering new files...
2024-06-18 13:43:44,756 [INFO] (BaseMetadataBackend) Compressed 0 existing files from 0.
2024-06-18 13:43:49,960 [INFO] (BaseMetadataBackend) Image processing statistics: {'total_processed': 612, 'skipped': {'already_exists': 0, 'metadata_missing': 0, 'not_found': 0, 'too_small': 0, 'other': 0}}
2024-06-18 13:43:49,964 [INFO] (BaseMetadataBackend) Enforcing minimum image size of 0.1. This could take a while for very-large datasets.
Enforcing minimum bucket size:   0%|                      | 0/6 [00:00<?, ?it/s]2024-06-18 13:43:49,966 [WARNING] (BaseMetadataBackend) Removed bucket 0.92 due to insufficient samples.
2024-06-18 13:43:49,966 [WARNING] (BaseMetadataBackend) Removed bucket 0.85 due to insufficient samples.
2024-06-18 13:43:49,966 [WARNING] (BaseMetadataBackend) Removed bucket 1.0 due to insufficient samples.
2024-06-18 13:43:49,966 [WARNING] (BaseMetadataBackend) Removed bucket 1.3 due to insufficient samples.
2024-06-18 13:43:49,966 [WARNING] (BaseMetadataBackend) Removed bucket 1.56 due to insufficient samples.
2024-06-18 13:43:49,967 [INFO] (BaseMetadataBackend) Completed aspect bucket update.
2024-06-18 13:43:49,970 [INFO] (DataBackendFactory) Configured backend: {'id': 'all_dataset', 'config': {'vae_cache_clear_each_epoch': False, 'probability': 1.0, 'repeats': 5, 'crop': False, 'crop_aspect': 'preserve', 'crop_aspect_buckets': None, 'crop_style': 'random', 'disable_validation': False, 'resolution': 0.5, 'resolution_type': 'area', 'caption_strategy': 'textfile', 'instance_data_root': '/workspace/input/dataset', 'maximum_image_size': 1.0, 'target_downsample_size': 0.55}, 'dataset_type': 'image', 'data_backend': <helpers.data_backend.local.LocalDataBackend object at 0x7fa33b42a6e0>, 'instance_data_root': '/workspace/input/dataset', 'metadata_backend': <helpers.metadata.backends.json.JsonMetadataBackend object at 0x7fa33b56d360>}
(Rank: 0)  | Bucket     | Image Count 
------------------------------
(Rank: 0)  | 0.79       | 567         
2024-06-18 13:43:49,971 [INFO] (DataBackendFactory) (id=all_dataset) Collecting captions.
Loading captions: 100%|█████████████████████████████████████████████████████████████████| 612/612 [00:00<00:00, 28887.82it/s]
2024-06-18 13:43:49,992 [INFO] (DataBackendFactory) (id=all_dataset) Initialise text embed pre-computation using the textfile caption strategy. We have 612 captions to process.
```

2. Repeats are ignored, I have "repeats" = 5 but it doesn't seem to influence the steps at all. Repeats could have helped to fill buckets, but they are just ignored.
```
[
    {
        "id": "all_dataset",
        "type": "local",
        "instance_data_dir": "/workspace/input/dataset",
        "crop": false,
        "crop_style": "random",
        "crop_aspect": "preserve",
        "resolution": 0.5,
        "resolution_type": "area",
        "minimum_image_size": 0.1,
        "maximum_image_size": 1.0,
        "target_downsample_size":0.55,
        "prepend_instance_prompt": false,
        "instance_prompt": null,
        "only_instance_prompt": false,
        "caption_strategy": "textfile",
        "cache_dir_vae": "/workspace/cache_images/",
        "vae_cache_clear_each_epoch": false,
        "probability": 1.0,
        "repeats": 5,
        "text_embeds": "alt-embed-cache",
        "skip_file_discovery": "",
        "preserve_data_backend_cache": true
    },
    {
        "id": "alt-embed-cache",
        "dataset_type": "text_embeds",
        "default": true,
        "type": "local",
        "cache_dir": "/workspace/cache_text_embeds/"
    }
]
```

3. VAECache warnings appeared up until the first save of a checkpoint, and then they disappeared.

**Comment by bghira** [comment]
you are correct, there is no option to force the trainer to reuse images in a single epoch. you simply have to ensure your math works out for dataset sizes.

**Comment by bghira** [comment]
repeats are how many times the entire dataset will be sampled in a given epoch. see the dataloader config page for info

**Comment by a-l-e-x-d-s-9** [comment]
In the dataloader config page:
```
repeats
Specifies the number of times all samples in the dataset are seen during an epoch. Useful for giving more impact to smaller datasets or maximizing the usage of VAE cache objects.
```
For 5 repeats the dataset should be repeated 5 times at the same epoch, just like it works in kohya, right?
It's not happening. There are no 5 times more steps per epoch.

**Comment by bghira** [comment]
but you only have a single dataset. what is the difference you expect to see between images being cycled 5 times in one epoch vs over 5 epochs? is it learning rate schedule related?

**Comment by a-l-e-x-d-s-9** [comment]
Sorry, but I don't understand why the "repeats" field in multidatabackend.json is ignored. From the [DATALOADER](https://github.com/bghira/SimpleTuner/blob/main/documentation/DATALOADER.md) page, it sounds exactly like the repeats field in the kohya trainer, and the example suggests it can be used in the same way.

If I specify repeats=5 for a dataset/folder, I assume that all images in this folder will be trained 5 times per epoch. In kohya, the shuffle of images and split into buckets happens once per epoch. For example, if I have a single image with an aspect ratio of 0.5 and use 5 repeats with a batch size of 5, it will allow the script to fill a single bucket of a ratio of 0.5 for training per epoch.

Currently, in SimpleTuner, is the repeats field ignored in a single dataset? Is this field supposed to function differently, or is there an issue with how it's being processed?

**Comment by a-l-e-x-d-s-9** [comment]
Thank you for the explanation. I know that captions are not shuffled.

I'm a little bit confused by your second sentence:
> the images are always sampled randomly and they are removed from the bucket as they are sampled.

This sounds like a contradiction to your first sentence:
> simpletuner doesn't do caption shuffling (ever) or --**changing buckets**-- by default, unless you have crop=true and crop_aspect=random.

Can you please clarify? If crop=false, and we have a bucket with 5 images, will this bucket be persistent with the same images over the course of the whole training of 100 epochs, or will this bucket be emptied and filled by other randomly shuffled images in the next epoch?

I am aware that the purpose of repeats is to balance between multiple subjects/datasets. But in my case, I used a single dataset with 5 repeats. In kohya, people use multiple repeats with a single folder all the time. I explained why such usage can be helpful to fill up buckets per epoch.

Regarding the crop options, it's very nice that you have them, but I think that using cropping options would not be good for my purposes.

I only tested repeats with a single dataset, crop=false. And seems that repeats are ignored in this case.

**Comment by bghira** [comment]
"changing buckets" = re-assigning the image to a different aspect ratio.

unless crop_aspect=random they are fixed values for all eternity, but in this case, the whole dataset maintains its same number of images per buckets, it's just that this configuration is then sampled for *n* times.

the reason it matters that images are removed from the bucket is because there are no longer enough images in the bucket to even sample it when you get down to an odd number. they are indexed by file path, and you can't have it in there twice.

**Comment by a-l-e-x-d-s-9** [comment]
I was thinking about buckets in two ways: 1. Group of all images with the same aspect ratio - like you think about them. 2. Groups of batch size that have the same aspect ratio - sub-division of 1 into batch size groups.
You are right, without cropping groups of aspect ratio will remain the same for the whole dataset.
But I still don't understand if the order of the training will be the same in all epochs. Or the images will be shuffled.
Let's say we have images 1,2,3,4 - with an aspect ratio 0.5, and image 5,6,7,8 with an aspect ratio 0.75.
With batch size 2 (for batch: <>)
You can train: <1,2>,<3,4>,<5,6>,<7,8>.
Is it possible that in the next epoch they will be reshuffled to be: <5,7>,<1,3>,<6,8>,<2,4> - aspect ratio maintained, but the order of batch size groups has changed, and images have been grouped differently in batches?

**Comment by bghira** [comment]
the order of sampling is different on each epoch, you can see this with the luminance chart on wandb never showing a repeating pattern.

this is because the random states change naturally over time, and we don't reset the training seed for each epoch.

**Comment by bghira** [comment]
next time please open a new discussion for your question instead of hijacking another one but your parquet file is likely incompatible with the format needed by ST. see the DATALOADER.md document

**Comment by a-l-e-x-d-s-9** [comment]
Ok, thank you for the information. Regarding the auto-deletion of images that are not filling a full bucket, it's especially a big problem for small and even medium-sized datasets due to the relatively high portion of the dataset that can be distributed between not full buckets. Automatic deletion of not full buckets can create a big distortion for such datasets due to the removal of a substantial portion of the desired training dataset. Trying to resolve buckets manually is possible, but in many cases, batch sizes are configured on the deployed training system at the last moment and adjusted due to other training parameters like optimizer, resolution, and many more, and restricted by the actual VRAM of the GPU used for training. It's very problematic to add a huge step of dataset editing in regard to batch size to try to fill the buckets for each adjustment of training parameters.

I think it should be handled by the training script. There are a few ways that other training scripts handle it. First of all, usage of repeats: if you have 5 repeats per epoch, all the buckets can be multiplied by 5, and possibly a lot of not full buckets can suddenly be filled. That's why I was trying to use repeats even with a single dataset. An additional way to handle not full buckets is to duplicate one of the images in the bucket. For example, if a bucket needs 5 images and has 4, we can duplicate one image to get 5, and each epoch, one of them will be selected randomly. The influence is that every image in the bucket will be trained on average 5% more, but the huge advantage is that they all will be trained, and the user doesn't need to think about or actively edit images to fill buckets all the time.

**Comment by twri** [comment]
Something like image_repeats in the dataloader could be useful in this case when you train on smaller datasets so you could better balance buckets as you do when you balance several data sources between each other with repeats.

**Comment by bghira** [comment]
it's not really the intention to be able to train precisely with 20 images in 6 different partially-filled buckets. i believe you are all correct, and that other tools work better for this. my trainer is used mostly for creating foundational models as research artifacts, and training on several million images.

the aspect bucket handling portion of the code, in every single trainer, is the most complicated part. the PixArt project's official trainer doesn't even support batched aspect ratio bucketing, it just forces you to a batch size of 1.

it's a delicate and complicated section of code, one that i've proudly put time into ensuring samples equally across the bucket lists without oversampling at all. this suggestion will lead to oversampling, on purpose.

if you want to crack open the sampler code and look at absolutely everything going on under the hood, i would like that - i'll accept and review pull requests, but it will take a long time to be approved as it's such a sensitive part of the code tree.

**Comment by bghira** [comment]
i think this was resolved through the calculation fix for #509.
